File: src/part2_multi_agent/cpp/ma_seirs.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <numeric>
#include <random>
#include <string>
#include <vector>

static constexpr int SUS = 0;
static constexpr int EXP = 1;
static constexpr int INF = 2;
static constexpr int REM = 3;

struct Params {
    int L = 300;
    int N = 20000;
    int T = 730;
    uint32_t seed = 12345;

    int init_S = 19980;
    int init_E = 0;
    int init_I = 20;
    int init_R = 0;

    double mean_dE = 3.0;
    double mean_dI = 7.0;
    double mean_dR = 365.0;

    double inf_force = 0.5;
};

static inline double neg_exp(std::mt19937 &gen, double mean) {
    // -mean * log(1 - U), U ~ Uniform[0,1)
    std::uniform_real_distribution<double> U(0.0, 1.0);
    double u = U(gen);
    return -mean * std::log(1.0 - u);
}

static inline int idx2d(int x, int y, int L) {
    return x * L + y;
}

static inline int wrap(int a, int L) {
    int r = a % L;
    return (r < 0) ? r + L : r;
}

static inline int neighborhood_I(const std::vector<int16_t> &Icount, int x, int y, int L) {
    // Moore (8 voisins) + cellule centrale, tore
    static const int dx[8] = {-1,-1,-1, 0, 0, 1, 1, 1};
    static const int dy[8] = {-1, 0, 1,-1, 1,-1, 0, 1};

    int total = Icount[idx2d(x,y,L)]; // centre inclus
    for (int k = 0; k < 8; ++k) {
        int xx = wrap(x + dx[k], L);
        int yy = wrap(y + dy[k], L);
        total += Icount[idx2d(xx,yy,L)];
    }
    return total;
}

static void run_one_sim(const Params &p, const std::string &out_csv) {
    std::mt19937 gen(p.seed);
    std::uniform_real_distribution<double> U01(0.0, 1.0);
    std::uniform_int_distribution<int> Upos(0, p.L - 1);

    // états
    std::vector<int8_t> state(p.N, SUS);
    // init exact
    int kS = p.init_S;
    int kE = p.init_E;
    int kI = p.init_I;
    int kR = p.init_R;

    int pos = 0;
    for (int i = 0; i < kS; ++i) state[pos++] = SUS;
    for (int i = 0; i < kE; ++i) state[pos++] = EXP;
    for (int i = 0; i < kI; ++i) state[pos++] = INF;
    for (int i = 0; i < kR; ++i) state[pos++] = REM;

    // mélanger
    std::shuffle(state.begin(), state.end(), gen);

    // durées individuelles fixes
    std::vector<double> dE(p.N), dI(p.N), dR(p.N);
    for (int i = 0; i < p.N; ++i) {
        dE[i] = neg_exp(gen, p.mean_dE);
        dI[i] = neg_exp(gen, p.mean_dI);
        dR[i] = neg_exp(gen, p.mean_dR);
    }

    // temps dans l'état
    std::vector<int16_t> t_in_state(p.N, 0);

    // positions
    std::vector<int16_t> x(p.N), y(p.N);
    for (int i = 0; i < p.N; ++i) {
        x[i] = static_cast<int16_t>(Upos(gen));
        y[i] = static_cast<int16_t>(Upos(gen));
    }

    // grille de comptage infectieux
    std::vector<int16_t> Icount(static_cast<size_t>(p.L) * p.L, 0);
    for (int i = 0; i < p.N; ++i) {
        if (state[i] == INF) {
            Icount[idx2d(x[i], y[i], p.L)]++;
        }
    }

    auto count_states = [&]() {
        int S=0,E=0,I=0,R=0;
        for (int i = 0; i < p.N; ++i) {
            switch (state[i]) {
                case SUS: S++; break;
                case EXP: E++; break;
                case INF: I++; break;
                case REM: R++; break;
            }
        }
        return std::array<int,4>{S,E,I,R};
    };

    // ordre aléatoire
    std::vector<int> order(p.N);
    std::iota(order.begin(), order.end(), 0);

    std::ofstream f(out_csv);
    if (!f) {
        throw std::runtime_error("Impossible d'ouvrir le fichier: " + out_csv);
    }
    f << "t,S,E,I,R\n";
    {
        auto c = count_states();
        f << 0 << "," << c[0] << "," << c[1] << "," << c[2] << "," << c[3] << "\n";
    }

    for (int t = 1; t <= p.T; ++t) {
        std::shuffle(order.begin(), order.end(), gen);

        for (int ii = 0; ii < p.N; ++ii) {
            int i = order[ii];

            int oldx = x[i], oldy = y[i];
            int nx = Upos(gen);
            int ny = Upos(gen);
            if (nx == oldx && ny == oldy) { // "autre cellule" (1 tentative)
                nx = Upos(gen);
                ny = Upos(gen);
            }

            // mise à jour Icount si infectieux et déplacement
            if (state[i] == INF && (nx != oldx || ny != oldy)) {
                Icount[idx2d(oldx, oldy, p.L)]--;
                Icount[idx2d(nx, ny, p.L)]++;
            }

            x[i] = static_cast<int16_t>(nx);
            y[i] = static_cast<int16_t>(ny);

            // temps discret (1 jour)
            t_in_state[i]++;

            int st = state[i];

            if (st == SUS) {
                int NI = neighborhood_I(Icount, nx, ny, p.L);
                if (NI > 0) {
                    double prob = 1.0 - std::exp(-p.inf_force * static_cast<double>(NI));
                    if (U01(gen) < prob) {
                        state[i] = EXP;
                        t_in_state[i] = 0;
                    }
                }
            } else if (st == EXP) {
                if (static_cast<double>(t_in_state[i]) > dE[i]) {
                    state[i] = INF;
                    t_in_state[i] = 0;
                    Icount[idx2d(nx, ny, p.L)]++; // devient infectieux
                }
            } else if (st == INF) {
                if (static_cast<double>(t_in_state[i]) > dI[i]) {
                    state[i] = REM;
                    t_in_state[i] = 0;
                    Icount[idx2d(nx, ny, p.L)]--; // quitte infectieux
                }
            } else if (st == REM) {
                if (static_cast<double>(t_in_state[i]) > dR[i]) {
                    state[i] = SUS;
                    t_in_state[i] = 0;
                }
            }
        }

        auto c = count_states();
        f << t << "," << c[0] << "," << c[1] << "," << c[2] << "," << c[3] << "\n";
    }

    f.close();
}

int main(int argc, char **argv) {
    Params p;
    std::string out = "data/part2_multi_agent/cpp_rep01.csv";

    // Arguments simples : --seed <int> --out <path> --T <int>
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--seed" && i + 1 < argc) {
            p.seed = static_cast<uint32_t>(std::stoul(argv[++i]));
        } else if (a == "--out" && i + 1 < argc) {
            out = argv[++i];
        } else if (a == "--T" && i + 1 < argc) {
            p.T = std::stoi(argv[++i]);
        } else {
            std::cerr << "Option inconnue: " << a << "\n";
            std::cerr << "Usage: " << argv[0] << " [--seed N] [--out path] [--T N]\n";
            return 1;
        }
    }

    try {
        run_one_sim(p, out);
        std::cout << "Terminé -> " << out << "\n";
    } catch (const std::exception &e) {
        std::cerr << "Erreur: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: src/part2_multi_agent/python/ma_seirs.py
#!/usr/bin/env python3
"""
Partie 2 — Modèle multi-agent SEIRS (CONFORME SUJET)
- N=20000 individus
- Grille 300x300 toroïdale
- Plusieurs agents par cellule
- Déplacement : à chaque pas, saut vers une cellule aléatoire (global)
- Voisinage : Moore (8 cases) + la case elle-même
- Infection : p = 1 - exp(-0.5 * N_I)
- Planification : ordre aléatoire, asynchrone, mise à jour immédiate
- Durées individuelles fixes tirées au début :
    dE ~ Exp(mean=3), dI ~ Exp(mean=7), dR ~ Exp(mean=365)
- 730 itérations (jours)
Sortie : CSV t,S,E,I,R
"""

from dataclasses import dataclass
from pathlib import Path
import argparse
import numpy as np


# États
SUS, EXP, INF, REM = 0, 1, 2, 3


@dataclass
class Params:
    L: int = 300          # grille LxL
    N: int = 20000        # nombre d'individus
    T: int = 730          # itérations (jours)
    seed: int = 12345     # graine principale

    # Initialisation imposée
    init_S: int = 19980
    init_I: int = 20
    init_E: int = 0
    init_R: int = 0

    # Moyennes exponentielles imposées
    mean_dE: float = 3.0
    mean_dI: float = 7.0
    mean_dR: float = 365.0

    # Coefficient infection imposé (0.5)
    inf_force: float = 0.5


# Offsets Moore (8 voisins)
MOORE = [(-1, -1), (-1, 0), (-1, 1),
         ( 0, -1),          ( 0, 1),
         ( 1, -1), ( 1, 0), ( 1, 1)]


def neg_exp(rng: np.random.Generator, mean: float) -> float:
    """
    Tirage exponentiel conforme à l'énoncé :
      -mean * log(1 - U), U ~ Uniform[0,1)
    """
    u = rng.random()
    return -mean * np.log(1.0 - u)


def init_population(p: Params):
    rng = np.random.default_rng(p.seed)

    # États init (exact)
    states = np.empty(p.N, dtype=np.int8)
    states[:p.init_S] = SUS
    states[p.init_S:p.init_S + p.init_E] = EXP
    states[p.init_S + p.init_E:p.init_S + p.init_E + p.init_I] = INF
    states[p.init_S + p.init_E + p.init_I:] = REM

    # Mélanger pour ne pas avoir les I tous regroupés
    rng.shuffle(states)

    # Durées individuelles (fixes pendant la simulation)
    dE = np.array([neg_exp(rng, p.mean_dE) for _ in range(p.N)], dtype=np.float64)
    dI = np.array([neg_exp(rng, p.mean_dI) for _ in range(p.N)], dtype=np.float64)
    dR = np.array([neg_exp(rng, p.mean_dR) for _ in range(p.N)], dtype=np.float64)

    # Temps écoulé dans l'état courant (initialement 0)
    t_in_state = np.zeros(p.N, dtype=np.int16)

    # Positions initiales aléatoires sur la grille (plusieurs agents par cellule autorisés)
    x = rng.integers(0, p.L, size=p.N, dtype=np.int16)
    y = rng.integers(0, p.L, size=p.N, dtype=np.int16)

    # Grille de comptage des infectieux par cellule (pour calculer N_I rapidement)
    Icount = np.zeros((p.L, p.L), dtype=np.int16)
    for i in range(p.N):
        if states[i] == INF:
            Icount[x[i], y[i]] += 1

    return rng, states, t_in_state, dE, dI, dR, x, y, Icount


def count_S_E_I_R(states: np.ndarray):
    S = int(np.sum(states == SUS))
    E = int(np.sum(states == EXP))
    I = int(np.sum(states == INF))
    R = int(np.sum(states == REM))
    return S, E, I, R


def neighborhood_I(Icount: np.ndarray, x: int, y: int, L: int) -> int:
    """
    N_I = infectieux dans Moore (8 voisins) + cellule centrale,
    sur une grille toroïdale.
    """
    total = int(Icount[x, y])  # cellule centrale incluse (exigé par le sujet)
    for dx, dy in MOORE:
        total += int(Icount[(x + dx) % L, (y + dy) % L])
    return total


def step_one_agent(i: int,
                   rng: np.random.Generator,
                   p: Params,
                   states: np.ndarray,
                   t_in_state: np.ndarray,
                   dE: np.ndarray, dI: np.ndarray, dR: np.ndarray,
                   x: np.ndarray, y: np.ndarray,
                   Icount: np.ndarray):
    """
    Mise à jour asynchrone d'un agent i :
    1) déplacement global aléatoire
    2) incrément temps dans l'état
    3) transitions (S->E probabiliste ; E->I ; I->R ; R->S)
    Mise à jour immédiate (Icount et états sont modifiés sur le champ).
    """
    L = p.L

    # ---------- 1) Déplacement global aléatoire (cellule choisie au hasard dans la grille)
    oldx, oldy = int(x[i]), int(y[i])
    nx = int(rng.integers(0, L))
    ny = int(rng.integers(0, L))

    # "vers une autre cellule" : on évite une fois le cas identique
    if nx == oldx and ny == oldy:
        nx = int(rng.integers(0, L))
        ny = int(rng.integers(0, L))

    # Mettre à jour Icount si l'agent est infectieux et change de cellule
    if states[i] == INF and (nx != oldx or ny != oldy):
        Icount[oldx, oldy] -= 1
        Icount[nx, ny] += 1

    x[i], y[i] = nx, ny

    # ---------- 2) temps écoulé dans l'état (discret, 1 pas = 1 jour)
    t_in_state[i] += 1

    # ---------- 3) transitions
    st = states[i]

    if st == SUS:
        NI = neighborhood_I(Icount, nx, ny, L)
        if NI > 0:
            # p = 1 - exp(-0.5 * N_I)
            prob = 1.0 - np.exp(-p.inf_force * NI)
            if rng.random() < prob:
                states[i] = EXP
                t_in_state[i] = 0  # reset temps dans l'état

    elif st == EXP:
        # devient I si temps écoulé > dE
        if float(t_in_state[i]) > float(dE[i]):
            states[i] = INF
            t_in_state[i] = 0
            # devient infectieux -> Icount +1 à la cellule actuelle
            Icount[nx, ny] += 1

    elif st == INF:
        if float(t_in_state[i]) > float(dI[i]):
            states[i] = REM
            t_in_state[i] = 0
            # quitte infectieux -> Icount -1
            Icount[nx, ny] -= 1

    elif st == REM:
        if float(t_in_state[i]) > float(dR[i]):
            states[i] = SUS
            t_in_state[i] = 0


def run_one_sim(p: Params, out_csv: Path):
    rng, states, t_in_state, dE, dI, dR, x, y, Icount = init_population(p)

    out_csv.parent.mkdir(parents=True, exist_ok=True)
    order = np.arange(p.N, dtype=np.int32)

    with out_csv.open("w", encoding="utf-8") as f:
        f.write("t,S,E,I,R\n")
        S, E, I, R = count_S_E_I_R(states)
        f.write(f"0,{S},{E},{I},{R}\n")

        for t in range(1, p.T + 1):
            rng.shuffle(order)  # planification aléatoire
            for i in order:
                step_one_agent(int(i), rng, p, states, t_in_state, dE, dI, dR, x, y, Icount)

            S, E, I, R = count_S_E_I_R(states)
            f.write(f"{t},{S},{E},{I},{R}\n")


def main():
    parser = argparse.ArgumentParser(description="SEIRS multi-agent (Partie 2)")
    parser.add_argument("--seed", type=int, default=12345, help="Graine RNG")
    parser.add_argument("--out", type=str, default="data/part2_multi_agent/python_rep01.csv",
                        help="Chemin du CSV de sortie")
    parser.add_argument("--T", type=int, default=730, help="Nombre d'itérations (jours)")

    args = parser.parse_args()

    p = Params(seed=args.seed, T=args.T)
    out = Path(args.out)
    run_one_sim(p, out)
    print("Terminé ->", out)


if __name__ == "__main__":
    main()
